src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt

```diff
@@ -84,3 +105,110 @@ fun getPolicyToTemplateMap(response: SearchResponse, xContentRegistry: NamedXCon
 @Suppress("UNCHECKED_CAST")
 fun <K, V> Map<K, V?>.filterNotNullValues(): Map<K, V> =
     filterValues { it != null } as Map<K, V>
+
+// get metadata from config index using doc id
+@Suppress("ReturnCount")
+suspend fun IndexMetadata.getManagedIndexMetaData(client: Client): ManagedIndexMetaData? {
+    try {
+        val getRequest = GetRequest(INDEX_MANAGEMENT_INDEX, indexUUID + "metadata")
+            .routing(this.indexUUID)
+        val getResponse: GetResponse = client.suspendUntil { get(getRequest, it) }
+        if (!getResponse.isExists || getResponse.isSourceEmpty) {
+            return null
+        }
+
+        return withContext(Dispatchers.IO) {
+            val xcp = XContentHelper.createParser(
+                NamedXContentRegistry.EMPTY,
+                LoggingDeprecationHandler.INSTANCE,
+                getResponse.sourceAsBytesRef, XContentType.JSON)
+            ManagedIndexMetaData.parseWithType(xcp, getResponse.id, getResponse.seqNo, getResponse.primaryTerm)
+        }
+    } catch (e: Exception) {
+        when (e) {
+            is IndexNotFoundException, is NoShardAvailableActionException -> {
+                log.error("Failed to get metadata because no index or shard not available")
+            }
+            else -> log.error("Failed to get metadata", e)
+        }
+
+        return null
+    }
+}
+
+/**
+ * multi-get metadata for indices
+ *
+ * @return list of metadata
+ */
+suspend fun Client.mgetManagedIndexMetadata(indices: List<Index>): List<ManagedIndexMetaData?> {
+    log.debug("trying to get back metadata for indices ${indices.map { it.name }}")
+
+    if (indices.isEmpty()) return emptyList()
+
+    val mgetRequest = MultiGetRequest()
+    indices.forEach {
+        mgetRequest.add(MultiGetRequest.Item(
+            INDEX_MANAGEMENT_INDEX, it.uuid + "metadata").routing(it.uuid))
+    }
+    var mgetMetadataList = mutableListOf<ManagedIndexMetaData?>()
+    try {
+        val response: MultiGetResponse = this.suspendUntil { multiGet(mgetRequest, it) }
+        mgetMetadataList = mgetResponseToList(response)
```

Comment:
Whats behavior when some of these fail and some succeed?

Reply:
extend the list type to be `List<Pair<ManagedIndexMetaData?, Exception?>?>`, failed response's exception will be saved and get processed when this list being used in change and retry API